synapse/lib/jsonstor.py

```diff
@@ -0,0 +1,380 @@
+        if refsbyts:
+            refs = s_msgpack.un(refsbyts)
+
+        refs += valu
+        if refs > 0:
+            self.slab.put(buid + b'refs', s_msgpack.en(refs), db=self.metadb)
+            return refs
+
+        # remove the meta entries
+        for lkey, byts in self.slab.scanByPref(buid, db=self.metadb):
+            self.slab.pop(lkey, db=self.metadb)
+
+        # remove the item data
+        self.slab.pop(buid, db=self.itemdb)
+        self.dirty.pop(buid, None)
+
+    async def setPathObj(self, path, item):
+        '''
+        Set (and/or reinitialize) the object at the given path.
+
+        NOTE: This will break any links by creating a new object.
+        '''
+        buid = os.urandom(16)
+
+        pkey = self._pathToPkey(path)
+
+        oldb = self.slab.replace(pkey, buid, db=self.pathdb)
+        if oldb is not None:
+            self._incRefObj(oldb, -1)
+
+        self.slab.put(buid + b'refs', s_msgpack.en(1), db=self.metadb)
+
+        self.dirty[buid] = item
+
+    async def getPathObj(self, path):
+        buid = self._pathToBuid(path)
+        if buid is None:
+            return None
+        return self._getBuidItem(buid)
+
+    def _getBuidItem(self, buid):
+        item = self.dirty.get(buid)
+        if item is not None:
+            return item
+
+        byts = self.slab.get(buid, db=self.itemdb)
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    def _pathToBuid(self, path):
+        pkey = self._pathToPkey(path)
+        return self.slab.get(pkey, db=self.pathdb)
+
+    async def delPathObj(self, path):
+        '''
+        Remove a path and decref the object it references.
+        '''
+        pkey = self._pathToPkey(path)
+        buid = self.slab.pop(pkey, db=self.pathdb)
+        if buid is not None:
+            self._incRefObj(buid, valu=-1)
+
+    async def setPathLink(self, srcpath, dstpath):
+        '''
+        Add a link from the given srcpath to the dstpath.
+        NOTE: This causes the item at dstpath to be incref'd
+        '''
+        srcpkey = self._pathToPkey(srcpath)
+        dstpkey = self._pathToPkey(dstpath)
+
+        if self.slab.get(srcpkey, db=self.pathdb):
+            raise s_exc.PathExists(path=srcpath)
+
+        buid = self.slab.get(dstpkey, db=self.pathdb)
+        if buid is None:
+            raise s_exc.NoSuchPath(path=dstpath)
+
+        self._incRefObj(buid, valu=1)
+        self.slab.put(srcpkey, buid, db=self.pathdb)
+
+    async def getPathObjProp(self, path, prop):
+
+        item = await self.getPathObj(path)
+        if item is None:
+            return None
+
+        for name in self._pathToTupl(prop):
+            item = item[name]
+
+        return item
+
+    def _pathToPkey(self, path):
+        path = self._pathToTupl(path)
+        return ('\x00'.join(path)).encode()
+
+    def _pathToTupl(self, path):
+
+        if isinstance(path, str):
+            path = path.split('/')
```

Comment:
Probably not a big deal but this returns a list, not a tuple which the name implies.